yew-macro/src/html_tree/html_element.rs

```diff
@@ -234,39 +234,53 @@ impl ToTokens for HtmlElement {
 
         let push_classes = match classes {
             Some(ClassesForm::Tuple(classes)) => {
+                let span = classes.span();
+                let classes: Vec<_> = classes.elems.iter().collect();
                 let n = classes.len();
                 let sr = stringify::stringify_at_runtime(quote! { __yew_classes });
-                Some(quote! {
-                    let mut __yew_classes = ::yew::virtual_dom::Classes::with_capacity(#n);
-                    #(__yew_classes.push(#classes);)*
+                Some(quote_spanned! {span=>
+                    {
+                        let mut __yew_classes = ::yew::html::Classes::with_capacity(#n);
+                        #(__yew_classes.push(#classes);)*
 
-                    if !__yew_classes.is_empty() {
-                        #vtag.__macro_push_attribute("class", #sr);
-                    } else {
-                        #vtag.__macro_push_attribute_placeholder("class");
-                    };
+                        if false {
+                            deprecated_use_of_class();
+                        }
+
+                        if !__yew_classes.is_empty() {
+                            #vtag.__macro_push_attribute("class", #sr);
+                        } else {
+                            #vtag.__macro_push_attribute_placeholder("class");
+                        };
+                    }
                 })
             }
             Some(ClassesForm::Single(classes)) => match classes.try_into_lit() {
                 Some(lit) => {
                     if lit.value().is_empty() {
                         None
                     } else {
+                        let span = classes.span();
                         let sr = lit.stringify();
-                        Some(quote! {
-                            #vtag.__macro_push_attribute("class", #sr);
+                        Some(quote_spanned! {span=>
+                            {
+                                #vtag.__macro_push_attribute("class", #sr);
+                            }
                         })
```

Comment:
I'm sorry if I caused a bit of confusion with my previous comment, but this doesn't need to be in its own expression block and it also doesn't need to use `quote_spanned!` because `#sr` already has a span attached to it.
```suggestion
                       Some(quote! {
                            #vtag.__macro_push_attribute("class", #sr);
                       })
```

You see, `quote_spanned!` doesn't overwrite the spans of the tokens you insert like `#this`, it only sets a span for the new tokens. In this case that would be `.__macro_push_attribute("class", _);`.
There's no way that an error should be associated with these tokens in the first place (unless it's our fault like if the method doesn't exist). You might think that it wouldn't hurt to attach a span to these anyway, but as I mentioned previously that will "turn on" Clippy which means we suddenly have to worry about that.
In the other snippets you modified we might start seeing warnings like `"__yew_classes" does not need to be mutable` because it now has a span attached to it.

This comment applies to the snippet above and below as well. One thing you do have to watch out for is the deprecation warning, which, as we know now, needs to have a span attached to it for it to be shown to the user.
I would handle that like this (added some comments for clarity):
```rust
Some(ClassesForm::Tuple(classes)) => {
    let classes = classes.elems.iter().collect::<Vec<_>>();
    let n = classes.len();
    let sr = stringify::stringify_at_runtime(quote! { __yew_classes });

    // notice the use of quote_spanned! here so that Clippy thinks `deprecated_use_of_class()` belongs to the input tokens.
    // I also moved the function definition in here because then we don't need to worry about the dead code warning
    // Technically we could also define the function using `quote!` instead of `quote_spanned!` to avoid the issue
    let deprecation_warning = quote_spanned! {classes.span()=>
        #[deprecated(
            note = "the use of `(...)` with the attribute `class` is deprecated and will be removed in version 0.19. Use the `classes!` macro instead."
        )]
        fn deprecated_use_of_class() {}

        if false {
            deprecated_use_of_class();
        };
    };

    // now we don't need to use `quote_spanned!` here which means we don't need to worry about Clippy warnings for the code we write in here
    Some(quote! {
        let mut __yew_classes = ::yew::virtual_dom::Classes::with_capacity(#n);
        #(__yew_classes.push(#classes);)*

        // insert the `deprecation_warning` here. It still has the span we attached above, but everything else here doesn't
        #deprecation_warning

        if !__yew_classes.is_empty() {
            #vtag.__macro_push_attribute("class", #sr);
        } else {
            #vtag.__macro_push_attribute_placeholder("class");
        };
    })
}
```

Reply:
Thanks a lot for all the explanation!! :heart: